### ISSUE TITLE
Async code testing paradigm is broken/ugly.

### ISSUE THREAD

**Issue by virl** [robolectric/robolectric#2977]
Robolectric have very broken and ugly async code testing paradigm, because it doesn't allow execution of implementation's hidden loopers in the test (the loopers that test have no direct references).

As the result, it is almost impossible to write a test for complex async code system that tests it as blackbox, without requiring the test to depend on all code's implementation details (specific loopers).

Robolectric should behave like Xcode's XCTest in that regard: allowing to unpause ALL loopers in the app with call of single method like .unpauseAllLoopers() and executing code on them in exactly same manner (on separate threads, etc.) as it behaves in real app.
(Frankly, this behaviour should be default and must not require even calling a method for this).

See #1993 #1994 #1727 #1711

**Comment by mikesol** [comment]
Thanks for doing this research!
As you reference a couple bugs I reported, I thought I'd add a workaround attempt that I've done with pretty good results.  I use cucumber to coordinate large-ish tests.  Aside from the benefits it brings in human readability, it also allows for things to be parallelized - a CI pipeline ships off artifacts to parallel testing servers that run a test in an isolated environment and merge the result down the pipeline.  This is, of course, a workaround, but at the end of the day it is a quite practical one that covers business-critical test cases and allows us to ship with confidence.

**Comment by virl** [comment]
@mikesol I think you're talking about different, but partially related problem.
This issue is about Robolectric's broken API for testing modern code due to Robolectric's fundamentally wrong multithreading architecture.

Your #3359 related to this in the sense that Robolectric's concurrency and scalability problems, being a different issue, have the same root in its wrong multithreading architecture that tries to "Freeze the World" in futile attempt to make totally deterministic the simplest of UI tests.

**Comment by emartynov** [comment]
Let me drop my 5 cents here. Below is IMO.

Using Robolectric for integration tests is probably a bad choice. The Robolectric is mainly used to test single class integration with Android system. In such tests, it is easy and reasonable to assume that multithreading could be replaced/tested by sequential execution runnables.

The real tests for the system integration and maybe e2e (if you prefer) are instrumental tests. no matter what they are hard to write, maintain and execute.

**Comment by virl** [comment]
@emartynov No, it is not so: even for single class multithreading is **implementation detail**, even if that class have only synchronous public API.
That is because class can (and will) accomplish intermediate parallel tasks to execute even synchronous request. Or that class can use arbitrary third-party library (which is implementation detail too that tests should not rely on) that will use multithreading.

"The Robolectric is mainly used to test single class integration with Android system. In such tests, it is easy and reasonable to assume that multithreading could be replaced/tested by sequential execution runnables"

Multithreading can never be replaced or tested by sequential execution, because nor JVM, nor Android makes such guarantees. JVM even have Java Memory Model that guarantees only happens-before execution barriers, not any deterministic behaviour between multiple threads without locks.

Therefore, replacing async testing with sequential runnables violates incapsulation of both tested code (tests will break when correct implementation changed to other correct implementation) and system API guarantees.

**Comment by deliciouslytyped** [comment]
Has there been any progress on this?
